libbeat/publisher/queue/memqueue/produce.go

```diff
@@ -42,66 +41,83 @@ type openState struct {
 	events chan pushRequest
 }
 
+// producerID stores the order of events within a single producer, so multiple
+// event acknowledgement callbacks can be coalesced into a single call.
+// It is defined as an explicit type to reduce cross-confusion with the id
+// of an event in the queue itself, which is a queue.EntryID.
+// with the
+type producerID uint64
+
 type produceState struct {
 	cb        ackHandler
-	dropCB    func(beat.Event)
+	dropCB    func(interface{})
 	cancelled bool
-	lastACK   uint32
+	lastACK   producerID
 }
 
 type ackHandler func(count int)
 
-func newProducer(b *broker, cb ackHandler, dropCB func(beat.Event), dropOnCancel bool) queue.Producer {
+func newProducer(b *broker, cb ackHandler, dropCB func(interface{}), dropOnCancel bool) queue.Producer {
 	openState := openState{
 		log:    b.logger,
 		done:   make(chan struct{}),
 		events: b.pushChan,
 	}
 
 	if cb != nil {
-		p := &ackProducer{broker: b, seq: 1, dropOnCancel: dropOnCancel, openState: openState}
+		p := &ackProducer{broker: b, dropOnCancel: dropOnCancel, openState: openState}
 		p.state.cb = cb
 		p.state.dropCB = dropCB
 		return p
 	}
 	return &forgetfulProducer{broker: b, openState: openState}
 }
 
-func (p *forgetfulProducer) Publish(event interface{}) bool {
-	return p.openState.publish(pushRequest{event: event})
+func (p *forgetfulProducer) makePushRequest(event interface{}) pushRequest {
+	resp := make(chan queue.EntryID, 1)
+	return pushRequest{
+		event: event,
+		resp:  resp}
+}
+
+func (p *forgetfulProducer) Publish(event interface{}) (queue.EntryID, bool) {
+	return p.openState.publish(p.makePushRequest(event))
 }
 
-func (p *forgetfulProducer) TryPublish(event interface{}) bool {
-	return p.openState.tryPublish(pushRequest{event: event})
+func (p *forgetfulProducer) TryPublish(event interface{}) (queue.EntryID, bool) {
+	return p.openState.tryPublish(p.makePushRequest(event))
 }
 
 func (p *forgetfulProducer) Cancel() int {
 	p.openState.Close()
 	return 0
 }
 
-func (p *ackProducer) Publish(event interface{}) bool {
-	return p.updSeq(p.openState.publish(p.makeRequest(event)))
+func (p *ackProducer) makePushRequest(event interface{}) pushRequest {
+	resp := make(chan queue.EntryID, 1)
+	return pushRequest{
+		event:    event,
+		producer: p,
+		// We add 1 to the id so the default lastACK of 0 is a
+		// valid initial state and 1 is the first real id.
+		producerID: producerID(p.producedCount + 1),
+		resp:       resp}
 }
 
-func (p *ackProducer) TryPublish(event interface{}) bool {
-	return p.updSeq(p.openState.tryPublish(p.makeRequest(event)))
-}
-
-func (p *ackProducer) updSeq(ok bool) bool {
-	if ok {
-		p.seq++
+func (p *ackProducer) Publish(event interface{}) (queue.EntryID, bool) {
+	id, published := p.openState.publish(p.makePushRequest(event))
+	if published {
+		p.producedCount++
```

Comment:
Is it thread safe to just increment the counter in both `Publish` and `TryPublish`?

Reply:
No, but neither is the code that this replaces. The queues enforce thread-safe access but a single producer is only supposed to be used by a single goroutine at a time (though now that you mention it I'm not sure the current shipper code respects that).

Reply:
(Thinking more: I think the shipper case is actually ok, because it doesn't use or need a producer callback, so it won't hit `ackProducer`, just `forgetfulProducer` which doesn't need to track anything.)